jobs: check liveness session when updating jobs

spaskob opened the pull request:
Currently when we resume a job, the jobs starts running on the local node
and periodically may update its progress. It may happen that the node gets
partitioned and that the job is adopted on another node. If the partitioned
node gets reconnected the first job may try to update its progress and
overwrite the second job's progress.

With the implementation of liveness leases we can avoid that by passing the
liveness session at the time of adoption to the job and the job `Update`
method can verify if the job liveness session has changed.

Release note: none.

Commit:
jobs: check liveness session when updating jobs

Currently when we resume a job, the jobs starts running on the local node
and periodically may update its progress. It may happen that the node gets
partitioned and that the job is adopted on another node. If the partitioned
node gets reconnected the first job may try to update its progress and
overwrite the second job's progress.

With the implementation of liveness leases we can avoid that by passing the
liveness session at the time of adoption to the job and the job `Update`
method can verify if the job liveness session has changed.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55120)
<!-- Reviewable:end -->

Commit:
jobs: deprecate functions needed for 20.1 only

As we improve the jobs code it turns out certain functions are used
only in test code or in the deprecated old lease adoption.

Release note: none.

Commit:
jobs: check for running or reverting in Set{Details,Progress}

Release note (bug fix):
If this check is missing concurrent requests to change the job's
state like pause and cancel will be overwritten. This on
occasion has made it hard to pause a running job since the
pause requests don't stay around long enough to be picked up
by the registry that owns the job.

ajwerner reviewed (approved):
> Reviewed 4 of 4 files at r1, 5 of 5 files at r2, 1 of 3 files at r3, 6 of 6 files at r4, 5 of 5 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55120)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/jobs/update.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/55120#-MJ7C6gHAknCMXaSr6ri:-MJ7C6gHAknCMXaSr6rj:bk0lsgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/340e6b8690a2ab939ba6b86352211a224e0a70b0/pkg/jobs/update.go#L109)):*
> ```Go
> 	if err := j.runInTxn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> 		selectStmt := "SELECT status, payload, progress FROM system.jobs WHERE id = $1"
> 		if j.sessionID != "" {
> ```

nit: might be cleaner to just always read the session ID and then return a more detailed error if the job's sessionID is non-zero and there's a mismatch. 


<!-- Sent from Reviewable.io -->

Commit:
jobs: refactor claim_session_id check

Always fetch the job only via the job id in `Update`
and only then if the `claim_session_id` is the same
as the on stored in the job objects. Report both
id-s if there is a mismatch.

Release note: none.

spaskob commented:
bors r+

spaskob commented:
bors r+
